### DISCUSSION TITLE
MK22FN512VLH12 Zephyr build error

### DISCUSSION THREAD

**Discussion by mondalaci** [zephyrproject-rtos/zephyr#63716] [Q&A] [answered]
Hi guys,

We need to port [our bare-metal, KSDK-based firmware](https://github.com/UltimateHackingKeyboard/firmware/tree/master/right/src) to Zephyr, but I can't even build the Hello World Zephyr sample application on the FRDM-K22F dev board:

```
Building hello_world
/bin/sh -c west build --build-dir /home/laci/download/hello_world/build /home/laci/download/hello_world --pristine --board frdm_k22f --no-sysbuild -- -DNCS_TOOLCHAIN_VERSION:STRING="NONE" -DBOARD_ROOT:STRING="/home/laci/projects/firmware-uhk" -DCONF_FILE:STRING="/home/laci/download/hello_world/prj.conf"

-- west build: generating a build system
Loading Zephyr default modules (Zephyr base).
-- Application: /home/laci/download/hello_world
-- CMake version: 3.25.1
-- Found Python3: /usr/bin/python3.10 (found suitable exact version "3.10.12") found components: Interpreter 
-- Cache files will be written to: /home/laci/.cache/zephyr
-- Zephyr version: 3.3.99 (/home/laci/apps/ncs/zephyr)
-- Found west (found suitable version "1.1.0", minimum required is "0.7.1")
-- Board: frdm_k22f
-- Found host-tools: zephyr 0.16.1 (/home/laci/apps/zephyr-sdk-0.16.1)
-- Found toolchain: zephyr 0.16.1 (/home/laci/apps/zephyr-sdk-0.16.1)
-- Found Dtc: /home/laci/apps/zephyr-sdk-0.16.1/sysroots/x86_64-pokysdk-linux/usr/bin/dtc (found suitable version "1.6.0", minimum required is "1.4.6") 
-- Found BOARD.dts: /home/laci/apps/ncs/zephyr/boards/arm/frdm_k22f/frdm_k22f.dts
In file included from /home/laci/apps/ncs/zephyr/boards/arm/frdm_k22f/frdm_k22f.dts:11,
                 from <command-line>:
/home/laci/apps/ncs/zephyr/boards/arm/frdm_k22f/frdm_k22f-pinctrl.dtsi:10:10: fatal error: nxp/kinetis/MK22FN512VLH12-pinctrl.h: No such file or directory
   10 | #include <nxp/kinetis/MK22FN512VLH12-pinctrl.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
CMake Error at /home/laci/apps/ncs/zephyr/cmake/modules/extensions.cmake:3714 (message):
  failed to preprocess devicetree files (error code 1):
  /home/laci/apps/ncs/zephyr/boards/arm/frdm_k22f/frdm_k22f.dts
Call Stack (most recent call first):
  /home/laci/apps/ncs/zephyr/cmake/modules/dts.cmake:223 (zephyr_dt_preprocess)
  /home/laci/apps/ncs/zephyr/cmake/modules/zephyr_default.cmake:115 (include)
  /home/laci/apps/ncs/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:66 (include)
  /home/laci/apps/ncs/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:92 (include_boilerplate)
  CMakeLists.txt:5 (find_package)


-- Configuring incomplete, errors occurred!
FATAL ERROR: command exited with status 1: /usr/bin/cmake -DWEST_PYTHON=/usr/bin/python3 -B/home/laci/download/hello_world/build -GNinja -DBOARD=frdm_k22f -DNCS_TOOLCHAIN_VERSION:STRING=NONE -DBOARD_ROOT:STRING=/home/laci/projects/firmware-uhk -DCONF_FILE:STRING=/home/laci/download/hello_world/prj.conf -S/home/laci/download/hello_world
```

As it turns out, the missing MK22FN512VLH12-pinctrl.h file is not part of the Zephyr repository, but it's part of the [https://github.com/zephyrproject-rtos/hal_nxp](https://github.com/zephyrproject-rtos/hal_nxp/tree/master/dts/nxp/kinetis) Zephyr module.

Strangely, the https://github.com/zephyrproject-rtos/zephyr/tree/main/modules/hal_nxp Zephyr module is already part of the main Zephyr repo, but it's almost empty compared to the above module.

The above makes me quite confused, and I need some help making Zephyr build on the MK22FN512VLH12.

I should also mention that I use Zephyr provided by the nRF Connect SDK, and I'm also using the nRF Connect for VS Code extension pack because we're also using nRF528* processors. I want to use the same environment to develop for both processors. According to my knowledge, this issue is not related to the Nordic tools.

Thanks in advance.

**Comment by nixward** [comment]
Hi, NCS doesn't include the NXP hal by default so you need to set up a west workspace and define your own west.yml manifest to allow the NXP hal see here for an example https://github.com/nrfconnect/ncs-example-application Once you have this set up and you run 'west update' you should see the nxp hal cloned and then you will be able to build.

**Reply by mondalaci** [reply]
Thanks for your help! I use a freestanding application. Looks like I have to use a workspace application according to https://docs.zephyrproject.org/latest/develop/application/index.html and https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/create_application.html

**Reply by mondalaci** [reply]
I've created a workspace application with the blinky project, but I don't know how to make it build for the `frdm_k22f` board. I assume I have to extend `west.yml` and `CMakeList.txt`, but I'm unsure how. Would you please help?

west.yml:
```
manifest:
  self:
    path: application
  remotes:
    - name: sdk
      url-base: https://github.com/nrfconnect
  projects:
    - name: sdk-nrf
      remote: sdk
      path: nrf
      revision: v2.4.99-dev2
      import:
        path-prefix: external
      clone-depth: 1
```

CMakeLists.txt:
```
# SPDX-License-Identifier: Apache-2.0

cmake_minimum_required(VERSION 3.20.0)
find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
project(blinky)

target_sources(app PRIVATE src/main.c)
```

**Reply by nordicjm** [reply]
As I said before, it will not work

**Reply by mondalaci** [reply]
Would you explain how to make it work with upstream Zephyr in terms of `west.yml`, `CMakeList.txt`, and possibly any other files, if needed, so that the compiler would find `nxp/kinetis/MK22FN512VLH12-pinctrl.h`?

**Reply by nordicjm** [reply]
You cannot

**Reply by mondalaci** [reply]
Are you saying I can't make Zephyr work with the MK22FN512VLH12? If so, why is the frdm_k22f board included?

**Reply by nordicjm** [reply]
Zephyr you can perfectly fine, zephyr with parts of ncs you cannot

**Reply by mondalaci** [reply]
I clearly explained in my previous comment that I'd use upstream Zephyr with NXP HAL.

I'd love to have instructions to achieve what I'm after, but I can't make progress when receiving incorrect/vague one-liners.

**Reply by nordicjm** [reply]
You have pasted a west manifest that is using ncs, ncs is NOT zephyr, for zephyr you need to use upstream zephyr repositories, have a look at the example-application manifest: https://github.com/zephyrproject-rtos/example-application/blob/main/west.yml
You would need to add `hal_nxp` to the allow list

**Comment by nordicjm** [comment]
> I should also mention that I use Zephyr provided by the nRF Connect SDK, and I'm also using the nRF Connect for VS Code extension pack because we're also using nRF528* processors. I want to use the same environment to develop for both processors. According to my knowledge, this issue is not related to the Nordic tools.

This is not allowed as per the license, if you want to use with non-nordic devices then you must use upstream zephyr without any part of the nRF connect SDK

**Reply by mondalaci** [reply]
I wasn't aware of this. I think the intention of Nordic is to profit from their intellectual property and, consequently, disallow the use of it for the silicon of another vendor. However, we're moving away from NXP to Nordic, so I doubt they'd ever care in this case. Strictly complying with their license would require different workspaces per build, complicating our build setup for no sensible reason.

**Reply by carlescufi** [reply]
That is up to you, but the license does state clearly that the code under the Nordic 5-clause license is only to be run on Nordic devices. As long as you ensure that this is always the case then you should be fine.

**Comment by DerekSnell** [comment]
Hi @mondalaci ,
The NXP board `frdm_k22f` is included in the Zephyr repo, and you can build Zephyr apps for that board.  I see in this discussion, you first started trying to use the Nordic tools for this.  And now you know that is not supported.

You can build with this board using CLI, or if you prefer VS Code, NXP offers our MCUXpresso extension.  I suggest you first get started following the steps for these options.  Either will include the hal_nxp repo in your West workspace.  That will provide a working example for you.  After that, you can modify the west.yml in your standalone application to include the required repos for your application.

To install Zephyr and the tools needed for NXP hardware, I recommend following the instructions for our [MCUXpresso extension for VS Code](https://github.com/nxp-mcuxpresso/vscode-for-mcux/wiki/).  These steps use the MCUXpresso Installer, which installs the Zephyr dependencies like West, Python, Zephyr SDK, etc., and it installs the drivers for the debug probes used with the NXP boards.  (Another option is to install manually following [Zephyr's Getting Started](https://docs.zephyrproject.org/latest/develop/getting_started/index.html), but the MCUXpresso Installer does simplify this process.)

Then you can follow the steps at [MCUXpresso extension for VS Code](https://github.com/nxp-mcuxpresso/vscode-for-mcux/wiki/) to import the Zephyr samples into VS Code, build, flash, and debug on the `frdm_k22f` board.  If you have any issues or questions using VS Code, please post them at the [vscode-for-mcux GitHub repo](https://github.com/nxp-mcuxpresso/vscode-for-mcux/issues).  Or you can do the same from CLI.  This is the build command I just tested:
```
west build -b frdm_k22f samples\basic\blinky --pristine
```
With the samples working for you, you can then update your standalone app to include these same repos.  Use the [west.yml](https://github.com/zephyrproject-rtos/zephyr/blob/main/west.yml) manifest included in the Zephyr repo as a reference.  At a minimum, your app's manifest will need to include the repos zephyr, hal_nxp, and cmsis.  As mentioned in this post, Zephyr's [example-application](https://github.com/zephyrproject-rtos/example-application/blob/main/west.yml) repo provides a great reference for trimming down the west.yml manifest to include only the repos required by your application.

Best regards

west.yml

**Reply by mondalaci** [reply, marked as answer]
Hi @DerekSnell,

Thank you very much for elaborating on the details!

I was able to build the Zephyr firmware not only for nRF528* but also for the K22, using the following `west.yml`:

```
manifest:
  self:
    path: uhk
  remotes:
    - name: sdk
      url-base: https://github.com/nrfconnect
    - name: zephyr
      url-base: https://github.com/zephyrproject-rtos

  projects:
    - name: sdk-nrf
      remote: sdk
      path: nrf
      revision: v2.4.2
      import:
        path-prefix: nrfconnect
      clone-depth: 1
    - name: hal_nxp
      remote: zephyr
      path: zephyr-hal-nxp
      clone-depth: 1
```

west has taken care of importing the `hal-nxp` CMake dependencies, so the `MK22FN512VLH12-pinctrl.h: No such file or directory` error message disappeared and the build succeeds.

Hopefully, I can make it from there. So far so good!
